text/0088-contract-level-messages.md

```diff
@@ -0,0 +1,237 @@
+///     * the call was made from an `Account` context. We only allow stored contracts to register messages.
+pub fn casper_manage_message_topic(
+        topic_name_ptr: *const u8,
+        topic_name_size: usize,
+        operation_ptr: *const u8,
+        operation_size: usize,
+    ) -> i32;
+```
+
+```rust
+/// Emits a new message on the specified topic.
+/// 
+/// The topic needs to have been already registered through a call to `casper_manage_message_topic` before emitting a message.
+/// 
+/// Will return an error if:
+///     * the message is not of a serialized [`MessagePayload`] type.
+///     * the length of the message is larger than the maximum length specified in the chainspec
+pub fn casper_emit_message(
+    topic_name_ptr: *const u8,
+    topic_name_size: usize,
+    message_ptr: *const u8,
+    message_size: usize,
+);
+```
+
+The `MessagePayload` enum initially allows for sending human readable strings as messages in the first iteration:
+```rust
+pub enum MessagePayload {
+    /// Human readable string message.
+    String(String),
+}
+```
+
+#### A new `Key` variant is introduced for indexing the message topic and message checksums:
+
+```rust
+pub enum Key {
+    ...
+    /// A `Key` under which a message topic or message checksum is stored.
+    Message(MessageAddr),
+}
+```
+
+`MessageAddr` is a composite key derived from the caller's entity address and a hash of the topic name:
+```rust
+pub struct MessageAddr {
+    /// The entity addr.
+    entity_addr: AddressableEntityHash,
+    /// A 32 byte BLAKE2b hash of the name of the message topic.
+    topic_hash: TopicNameHash,
+    /// The message index. If the message index is `None`, the `MessageAddr` points to the topic control record.
+    message_index: Option<u32>,
+}
+```
+
+There are 2 types of stored values that are written to global state under this key.
+
+The control record for the message topic:
+```rust
+pub struct MessageTopicSummary {
+    /// Number of messages in this topic.
+    pub(crate) message_count: u32,
+    /// Block timestamp in which these messages were emitted.
+    pub(crate) blocktime: BlockTime,
+}
+```
+
+
+Checksum of the message payload that was emitted:
+```rust
+pub struct MessageChecksum(
+    pub  [u8; 32], //a 32 byte BLAKE2b hash of the serialized `MessagePayload` that was emitted.
+);
+```
+
+When a new message is emitted on a specified topic the `message_count` in the `MessageTopicSummary` record is incremented.
+Messages have ephemeral nature by design. In order to not increase the size of global state in an uncontrolled manner, the message count is reset to zero with every newly emitted message that belongs to a new block and the block time is updated as well. The old message checksum entries are pruned from global state.
+Of-chain applications can still query the old message checksums if required by using the state root hash of the block where the messages were emitted.
```

Comment:
```suggestion
Off-chain applications can still query the old message checksums if required by using the state root hash of the block where the messages were emitted.
```

Reply:
Fixed in: https://github.com/casper-network/ceps/pull/88/commits/5ed76edf97bdf5e31cada98ce01522dad3e109d5